odoo/models.py

```diff
@@ -774,6 +769,23 @@ def splittor(rs):
                         else:
                             current[i] = False
 
+        # if any xid should be exported, only do so at toplevel
+        if batch_invalidate and any(f[-1] == 'id' for f in fields):
+            bymodels = collections.defaultdict(set)
+            xidmap = {}
+            # collect all the tuples in "lines" (along with their coordinates)
+            for i, line in enumerate(lines):
+                for j, cell in enumerate(line):
+                    if type(cell) is tuple:
+                        bymodels[cell[0]].add(cell[1])
+                        xidmap[cell] = (i, j)
```

Comment:
What if several cells contain the same `(model_name, record_id)`?
I think that `xidmap` values should be lists of pairs `(i, j)`.

Reply:
BTW, wouldn't it be simpler to index `xidmap` by `record`?
And you could put `record` in the cell as well, instead of the pair `(record._name, record.id)`.

Reply:
> What if several cells contain the same (model_name, record_id)?
> I think that xidmap values should be lists of pairs (i, j).

Yup looks like I missed this bit.

> BTW, wouldn't it be simpler to index xidmap by record?

Not really, and since we can likely have a large number of placeholders I would assume tuples are somewhat lighter than records.